atc/exec/set_pipeline_step.go

```diff
@@ -157,13 +165,17 @@ func (step *SetPipelineStep) run(ctx context.Context, state RunState) error {
 		logger.Debug("no-diff")
 
 		fmt.Fprintf(stdout, "no diff found.\n")
+		err := pipeline.SetParentIDs(step.metadata.JobID, step.metadata.BuildID)
```

Comment:
Hmm, I'm not sure this is the best approach as I don't think it'll handle concurrent updates very well, given that we only want to set the parent IDs if the build is "newer" (i.e. has a larger `BuildID`).

Right now, `build.SavePipeline` will ensure that the parent IDs only get updated if the build is newer than the previous parent, and returns `ErrSetByNewerBuild` if not (the `set_pipeline` step should just log a warning if it receives this error).

So to me it seems like we should always be calling `build.SavePipeline`, rather than adding this extra `SetParentIDs` method. Unless you came up with a different approach for handling concurrent `set_pipeline`s?

Reply:
That's a good point, we'll have to add a check for that.

I guess another question is if it's possible for a pipeline to be set by 2 different jobs (from the same or even different pipelines). There's nothing in the RFC that forbids this but I'm not sure if there's a use case in the wild.

Reply:
Yeah - we discussed this with @vito, and his thoughts were that that's a use case we probably shouldn't try to support (though should document what happens in that case).

If we follow the same "only update if the build ID is larger" approach, it shouldn't matter much if different jobs are setting the same pipeline, since build IDs are auto incrementing across all jobs.

Reply:
One possible issue is that we "lose" previous build history if another job configures the same pipeline:
```
job1 kicks off 2 builds, job 2 kicks off 1 build

job1 build2 finishes -> initial create
job2 build1 finishes -> overwrites previous config
job1 build1 finishes -> overwrites previous config
```

Reply:
Oh right, forgot that build_id is globally incremental

Reply:
With the increasing build ID approach, the idea is that the resultant pipeline config is at least always independent of the order the builds finish in - it depends on the order the builds were created in

Reply:
@aoldershaw we could change how `set_pipeline` works where it always calls `build.SavePipeline` regardless if there's a diff or not. I think that would be fine. The main benefit we have from calling `SetParentIDs` instead of `SavePipeline` in this scenario, where there's no diff, is less load against the database. We only have to make one query instead of 9+ queries (it's actually almost always more than nine because we issue one query per resource/resource type/job/etc.). I'm trying to be conscious of when we issue more queries than is actually necessary; I think this is one of those situations where we can stay ahead performance-wise.

To solve the race-condition you mentioned, we can add `WHERE parent_build_id < buildID` to the query in `SetParentIDs`, like we did in `SavePipeline` and return the same `ErrSetByNewerBuild`.

Reply:
@taylorsilva although, it's only 9+ queries if the pipeline will be set (i.e. the `buildID` is newer) - and in that case, we'd need to make those queries anyway.

So with `SetParentIDs`, we're making 10+ queries in the common case rather than 9+. I think the stronger argument for only updating in `build.SavePipeline` is so the update is all-or-nothing, rather than introducing the potential for inconsistent state in the DB by having 2 updates (what if the `SetParentIDs` update succeeds but the `SavePipeline` one fails? the `parent_build_id` and `parent_job_id` are no longer accurate)

Reply:
I feel like I'm not understanding something that you're trying to convey. Here's my response based on this assumption: you think we run both `SetParentIDs` and `SavePipeline` every time. Let me know if that assumption is wrong. The below may be invalid.

---

We would never run both `SetParentIDs` AND `SavePipeline`. It's one OR the other based on whether a diff exists or not.

If you look at `set_pipeline` currently, it exits early when there's no dif:
https://github.com/concourse/concourse/blob/becd8ab072ccaa6cdb513e73dd8e7dbe69c384f2/atc/exec/set_pipeline_step.go#L193-L200

We're trying to maintain that behaviour. When we fall into this `if` we do the minimum that we expect from this step which is updating the job and build id of the pipeline the step sets.

Reply:
Ah, your assumption is correct, thanks for the clarification. I didn't read the code carefully enough. Adding `WHERE parent_build_id < buildID` to `SetParentIDs` makes sense!